File: ydb/benches/sdk_compare/metrics.rs
use std::time::Duration;

use anyhow::{Context, Result};
use hdrhistogram::Histogram;
use serde::Serialize;

const LOWEST_LATENCY_US: u64 = 1;
const HIGHEST_LATENCY_US: u64 = 300_000_000;
const SIGNIFICANT_DIGITS: u8 = 3;

pub(crate) struct LatencyRecorder {
    histogram: Histogram<u64>,
}

impl LatencyRecorder {
    pub(crate) fn new() -> Result<Self> {
        let histogram =
            Histogram::new_with_bounds(LOWEST_LATENCY_US, HIGHEST_LATENCY_US, SIGNIFICANT_DIGITS)
                .context("failed to create latency histogram")?;
        Ok(Self { histogram })
    }

    pub(crate) fn record(&mut self, latency: Duration) -> Result<()> {
        let latency_us = latency.as_micros().max(1);
        let latency_us = u64::try_from(latency_us).context("latency does not fit into u64")?;
        self.histogram
            .record(latency_us)
            .with_context(|| format!("latency {latency_us} us is outside histogram bounds"))
    }

    pub(crate) fn merge(&mut self, other: &Self) -> Result<()> {
        self.histogram
            .add(&other.histogram)
            .context("failed to merge latency histograms")
    }

    pub(crate) fn count(&self) -> u64 {
        self.histogram.len()
    }

    pub(crate) fn summary(&self) -> LatencyMetric {
        let count = self.count();
        let latency_us = (count > 0).then(|| LatencySummary {
            min: self.histogram.min(),
            max: self.histogram.max(),
            mean: self.histogram.mean(),
            p50: self.histogram.value_at_quantile(0.5),
            p95: self.histogram.value_at_quantile(0.95),
            p99: self.histogram.value_at_quantile(0.99),
            p99_9: self.histogram.value_at_quantile(0.999),
        });

        LatencyMetric { count, latency_us }
    }
}

#[derive(Debug, Serialize)]
pub(crate) struct LatencyMetric {
    pub(crate) count: u64,
    pub(crate) latency_us: Option<LatencySummary>,
}

#[derive(Debug, Serialize)]
pub(crate) struct LatencySummary {
    pub(crate) min: u64,
    pub(crate) max: u64,
    pub(crate) mean: f64,
    pub(crate) p50: u64,
    pub(crate) p95: u64,
    pub(crate) p99: u64,
    pub(crate) p99_9: u64,
}

File: benchmarks/sdk-compare/cpp/src/core/metrics.cpp
#include "metrics.h"

#include <algorithm>
#include <stdexcept>

namespace sdk_compare {
namespace {

inline constexpr std::int64_t lowest_latency_us = 1;
inline constexpr std::int64_t highest_latency_us = 300'000'000;
inline constexpr int significant_digits = 3;

std::uint64_t as_unsigned(std::int64_t value) {
    if (value < 0) {
        throw std::runtime_error("latency histogram returned a negative value");
    }
    return static_cast<std::uint64_t>(value);
}

}  // namespace

void to_json(nlohmann::json& value, const LatencyMetric& metric) {
    value = nlohmann::json{
        {"count", metric.count},
        {"latency_us",
         metric.latency_us ? nlohmann::json(*metric.latency_us)
                           : nlohmann::json(nullptr)},
    };
}

LatencyRecorder::LatencyRecorder() {
    hdr_histogram* histogram = nullptr;
    if (hdr_init(lowest_latency_us,
                 highest_latency_us,
                 significant_digits,
                 &histogram) != 0) {
        throw std::runtime_error("failed to create latency histogram");
    }
    histogram_.reset(histogram);
}

void LatencyRecorder::HistogramDeleter::operator()(
    hdr_histogram* histogram) const {
    hdr_close(histogram);
}

void LatencyRecorder::record(std::chrono::steady_clock::duration latency) {
    const auto micros =
        std::chrono::duration_cast<std::chrono::microseconds>(latency).count();
    const auto value = std::max<std::int64_t>(micros, lowest_latency_us);
    if (!hdr_record_value(histogram_.get(), value)) {
        throw std::runtime_error("latency " + std::to_string(value) +
                                 " us is outside histogram bounds");
    }
}

void LatencyRecorder::merge(const LatencyRecorder& other) {
    if (this == &other) {
        throw std::runtime_error(
            "cannot merge a latency histogram into itself");
    }

    const auto dropped = hdr_add(histogram_.get(), other.histogram_.get());
    if (dropped != 0) {
        throw std::runtime_error(
            "failed to merge latency histograms: " + std::to_string(dropped) +
            " values were dropped");
    }
}

LatencyMetric LatencyRecorder::summary() const {
    const auto count = as_unsigned(histogram_->total_count);
    if (count == 0) {
        return LatencyMetric{
            .count = count,
            .latency_us = std::nullopt,
        };
    }
    return LatencyMetric{
        .count = count,
        .latency_us =
            LatencySummary{
                .min = as_unsigned(hdr_min(histogram_.get())),
                .max = as_unsigned(hdr_max(histogram_.get())),
                .mean = hdr_mean(histogram_.get()),
                .p50 = as_unsigned(
                    hdr_value_at_percentile(histogram_.get(), 50.0)),
                .p95 = as_unsigned(
                    hdr_value_at_percentile(histogram_.get(), 95.0)),
                .p99 = as_unsigned(
                    hdr_value_at_percentile(histogram_.get(), 99.0)),
                .p99_9 = as_unsigned(
                    hdr_value_at_percentile(histogram_.get(), 99.9)),
            },
    };
}

}  // namespace sdk_compare
